File: 117_PopulatingNextPointersII.java
class PopulatingNextPointersII extends PopulatingNextPointers {
  // BFS is the same
  // Thus only DFS (recursive) approach
  public Node connect(Node root) {
    dfs(root);
    return root;
  }

  public void dfs(Node node) {
    if (node==null || (node.left==null && node.right==null)) return;

    if (node.left!=null && node.right!=null) 
      node.left.next = node.right;
    Node p = node.right!=null?node.right:node.left;
    Node q = node.next;

    // ....
    while (q!=null && q.left==null && q.right==null)
      q = q.next;
    
    p.next = q!=null?(q.left!=null?q.left:node.next.right):null;

    // ....
    dfs(node.right);
    dfs(node.left);
  }
}

File: 117_PopulatingNextPointersII.cpp
#define PERFECT_BINARY_TREE_NODE
#include "DataStructure.h"

class Solution {
  private:
    void dfs(Node* root) {
      if (!root) return;

      Node* p;
      if (root->left && root->right) {
        root->left->next = root->right;
        p = root->right;
      }
      else
        p = root->left ? root->left : root->right;

      if (!p) return;

      Node* q = root->next;
      while (q && !q->left && !q->right)
        q = q->next;

      p->next = q ? (q->left ? q->left : q->right) : nullptr;

      dfs(root->right);
      dfs(root->left);
    }

  public:
    Node* connect(Node* root) {
      dfs(root);
      return root;
    }
};

